client/src/App.tsx

```diff
@@ -1,60 +1,61 @@
+import validatePlayerName from "./validation/validate_player_name";
 import {
-  chooseRandomStat,
   calculateRoundWinner,
-  updateRoundWinners,
-  updatePlayerScores,
+  chooseRandomStat,
   updatePlayerCards,
   updatePlayerIsCardShown,
   updatePlayerIsCardShownAll,
+  updatePlayerScores,
+  updateRoundWinners,
 } from "./service/round/round";
-
-//Game starts with two players each with five cards
-//Clicking "Play" triggers startRound to start the first round (and the game)
-//showPlayerCard reveals the top card in the current player's hand
-
-//If the player !isHuman there is a setTimeout to mimic a human choosing a stat
-//chooseRandomStat selects a random stat from the current player's card
-//playRound then runs with the chosen stat
-
-//If the player isHuman there should be a prompt to pick a stat from the card i.e. "Please chose a stat player 2"
-//The player clicks on a stat which triggers playRound to run with their chosen stat
-
-//showAllCards reveals all of the player's top cards
-//A setTimeout is used to allow time for all the cards to be shown
-//calculateRoundWinner then compares the value of chosen stat with the value of that stat on the top card of every other player
-//The player with the card with the highest value for that characteristic wins the round
-//If there is a draw between multiple highest values then the current player wins the round
-
-//endRound then runs to end the round and prepare for the next round
-//hideAllCards turns all the cards face down
-//A setTimeout is used to allow time for all of the cards to be hidden
-//The top card, i.e. the card used this round, for each player is set to the bottom of their hand
-
-//If the currentRoundRef === game.totalRounds then the game ends and a winner is declared
-//If the curerntRoundRef !== game.totalRounds then startRound() runs and to start the round for the next player
+import { Stat } from "./types/card/card.types";
 
 function App() {
   const [game, setGame] = useState<Game>(
     initialiseGame(DEFAULT_PLAYERS, DEFAULT_ROUNDS)
   );
   const player1 = game.players[0];
   const player2 = game.players[1];
-
   const { players, totalRounds, roundWinners } = game;
 
+  const [inputPlayerNames, setInputPlayerNames] = useState<PlayerNameAndId[]>(
+    getInitialPlayerNames(game.players)
+  );
+
+  const updatePlayerName = (id: string, value: string) => {
+    setInputPlayerNames((currentInput) => {
+      return currentInput.map((input) =>
+        input.id === id ? { id: id, name: value } : input
+      );
+    });
+
+    const errorMessages = validatePlayerName(value);
+
+    if (errorMessages.length === 0) {
+      setGame((currentGame) => {
+        const updatedPlayers = currentGame.players.map((player) => {
+          return player.id === id ? { ...player, name: value.trim() } : player;
+        });
+        return {
+          ...currentGame,
+          players: updatedPlayers,
+        };
+      });
+    }
+  };
```

Comment:
Out of interest could all this login happen in PlayerScore - like the update and validation? Then there's less in App.tsx?

Reply:
Nice let me know if you want me to review after.

Reply:
@jennysivapalan I've moved all the validation logic into PlayerScore, it also now has the separate state for the input name before setting it in the game state, good idea on that one. The rest of updatePlayerName I'm pretty sure needs to stay in App.tsx as it's calling setGame to update the game state?

Reply:
Makes sense